### PR TITLE
lint: add guards to prevent the use of std::any/optional/variant

### PR DESCRIPTION
**Pull request by stedsome**
Signed-off-by: Yifan Yang <needyyang@google.com>

Per the discussion on https://github.com/envoyproxy/envoy/issues/12341, use of these standard library functions are temporarily banned to prevent runtime failure when running in iOS11.   

Commit Message:
Additional Description:
Risk Level:
Testing: Low. As this PR is designed to prevent an issue from happening and to date, there are no use of these symbols in the Envoy codebase. 
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: tools/code_format/check_format.py [hunk 1]
```diff
@@ -660,6 +660,14 @@ def checkSourceLine(line, file_path, reportError):
     # The std::atomic_* free functions are functionally equivalent to calling
     # operations on std::atomic<T> objects, so prefer to use that instead.
     reportError("Don't use free std::atomic_* functions, use std::atomic<T> members instead.")
+  # Blocking the use of std::any, std::optional, std::variant for now as iOS 11 does not
```
Comment (junr03) [inline tools/code_format/check_format.py:663]: nit: also macOS 10.13

File: tools/testdata/check_format/std_any.cc [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#include <any>
+
+namespace Envoy {
+
+std::any foo;
+
+} // namespace Envoy
```
Comment (junr03) [inline tools/testdata/check_format/std_any.cc:7]: nit: newline

**Comment by stedsome** [comment]
should be fixed now. @junr03

**Comment by junr03** [comment]
@lizan @ggreenway fyi

**Pull request merged**